components/tikv_util/src/resizable_threadpool.rs

```diff
@@ -1,161 +1,340 @@
 // Copyright 2021 TiKV Project Authors. Licensed under Apache-2.0.
 
-use std::sync::{Arc, RwLock};
+use std::sync::{Arc, Mutex, Weak};
 
 use futures::Future;
-use tokio::{io::Result as TokioResult, runtime::Runtime};
-
-/// DaemonRuntime is a "background" runtime, which contains "daemon" tasks:
-/// any task spawn into it would run until finish even the runtime isn't
-/// referenced.
-pub struct DaemonRuntime(Option<Runtime>);
-
-impl DaemonRuntime {
-    /// spawn a daemon task to the runtime.
-    pub fn spawn(self: &Arc<Self>, f: impl Future<Output = ()> + Send + 'static) {
-        let wkr = self.clone();
-        self.0.as_ref().unwrap().spawn(async move {
-            f.await;
-            drop(wkr)
-        });
-    }
+use tokio::{
+    io::Result as TokioResult,
+    runtime::{Builder, Handle, Runtime},
+};
+use tokio_util::task::task_tracker::TaskTracker;
+
+struct DeamonRuntime {
+    inner: Option<Runtime>,
+    tracker: TaskTracker,
+}
 
-    /// create a daemon runtime from some runtime.
-    pub fn from_runtime(rt: Runtime) -> Arc<Self> {
-        Arc::new(Self(Some(rt)))
+impl DeamonRuntime {
+    pub fn spawn<Fut>(&self, fut: Fut)
+    where
+        Fut: Future<Output = ()> + Send + 'static,
+    {
+        self.inner
+            .as_ref()
+            .unwrap()
+            .spawn(self.tracker.track_future(fut));
     }
 }
 
-impl Drop for DaemonRuntime {
+impl Drop for DeamonRuntime {
     fn drop(&mut self) {
-        if let Some(runtime) = self.0.take() {
+        if let Some(runtime) = self.inner.take() {
             runtime.shutdown_background();
         }
     }
 }
 
+#[derive(Clone)]
+pub struct DeamonRuntimeHandle {
+    inner: Weak<Mutex<DeamonRuntime>>,
```

Comment:
`Runtime::spawn` takes a `&self`, `TaskTracker::track_future` also takes a `&self`.

There is no need to use a Mutex since nothing requires a unique reference.

```suggestion
    inner: Weak<DeamonRuntime>,
```

Reply:
The mutex is to make sure the underlying runtime would always be the after `adjust_with` one